### ISSUE TITLE
Error in enumerate(data_loader) 

### ISSUE THREAD

**Issue by Bai426** [open-mmlab/mmdetection#2499]
Hi, I meet an error which is quite weird.
I used my own dataset, including 2000 images, but after being loaded by data_loader, the model only trained several images instead of these 2000 ones.

So I change a place in runner.py like this:
a) The original code in line 261 for def train():  
     **for i, data_batch in enumerate(data_loader):**
b) I used the following code to replace it:
    **for i in range(len(data_loader)):                     
          data_batch = next(iter(data_loader))** 
c) at last, it can load all the images and train.

My question is why 'enumerate' cannot function here. Anyone can explain this? Does this change bring any other issues?
 
**Checklist**
1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**

**Reproduction**
1. What command or script did you run?
```
A placeholder for the command.
```
2. Did you make any modifications on the code or config? Did you understand what you have modified?
3. What dataset did you use?

**Environment**

1. Please run `python mmdet/utils/collect_env.py` to collect necessary environment infomation and paste it here.
2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source]
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error trackback here.
```
A placeholder for trackback.
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

**Comment by PanJianning** [comment]
* Does this change brings any other issues?
  You should not call `iter` in for loop, otherwise you will always get the first batch if the dataloader are not shuffled.
 
And I think enumerate has nothing wrong.

**Comment by Bai426** [comment]
Until now I didn't have an issue with training.
I have a little worry about that the batch-data are not shuffled in this case, because the function enumerating can shuffle the data.

**Comment by hellock** [comment]
This is not a usual usage case. As @PanJianning suggested, it obtains only the first batch.
```python
for i in range(len(data_loader)):
    data_batch = next(iter(data_loader))
```

**Comment by Bai426** [comment]
No, it can get all the images, because they have already included in dataloader, and my training is quite well.

Sent from Outlook Mobile<https://aka.ms/blhgte>

________________________________

**Comment by PanJianning** [comment]
Every time you call `iter`,  the dataloader is shuffled again, so you can get different images in the first batch. But that does not guarantee you can get all the images.
